core-bundle/src/DependencyInjection/Compiler/RegisterFragmentsPass.php

```diff
@@ -44,10 +51,100 @@ public function process(ContainerBuilder $container): void
+        $annotationReader = new AnnotationReader(new DocParser());
+
+        foreach ($dirs as $dir => $namespace) {
+            $container->addResource(new DirectoryResource($dir, '/\.php$/'));
+
+            $finder = Finder::create()->files()->in($dir)->name('/\.php$/');
+
+            /** @var SplFileInfo $file */
+            foreach ($finder as $file) {
+                $class = $namespace . '\\' . \pathinfo($file->getFilename(), PATHINFO_FILENAME);
+
+                if (!class_exists($class)) {
+                    continue;
+                }
+
+                $reflection = new \ReflectionClass($class);
+
+                foreach ($annotationReader->getClassAnnotations($reflection) as $annotation) {
+                    if (!$annotation instanceof ContentElement || !$annotation instanceof FrontendModule) {
+                        continue;
+                    }
+
+                    var_dump($class, $annotation);
+                    $serviceId = $annotation->service ?: $class;
+
+                    if ($container->hasDefinition($serviceId)) {
+                        $definition = $container->getDefinition($serviceId);
+                    } else {
+                        $definition = new Definition($class);
+                        $container->setDefinition($serviceId, $definition);
+                    }
+
+                    $definition->setPublic(true);
+
+                    $attributes = [
+                        'category' => $annotation->category,
+                        'method' => '__invoke', // TODO
+                        //'options' => $annotation->options, // TODO
+                        'renderer' => $annotation->renderer,
+                        'template' => $annotation->template,
+                        'type' => $annotation->type,
+                    ];
+
+                    $attributes['type'] = $this->getFragmentType($definition, $attributes);
+
+                    $tag = ($annotation instanceof ContentElement) ? ContentElementReference::TAG_NAME : FrontendModuleReference::TAG_NAME;
+
+                    $identifier = sprintf('%s.%s', $tag, $attributes['type']);
+                    $reference = new Reference($serviceId);
+                    $config = $this->getFragmentConfig($container, $reference, $attributes);
+
+                    if (is_a($definition->getClass(), FragmentPreHandlerInterface::class, true)) {
+                        $preHandlers[$identifier] = $reference;
+                    }
+
+                    if (is_a($definition->getClass(), FragmentOptionsAwareInterface::class, true)) {
+                        $definition->addMethodCall('setFragmentOptions', [$attributes]);
+                    }
+
+                    $registry->addMethodCall('add', [$identifier, $config]);
+                }
+            }
+        }
+
+        $this->addPreHandlers($container, $preHandlers);
+    }
+
+    private function getAnnotatedControllersDirs(ContainerBuilder $container)
```

Comment:
Not sure about this. How does Symfony do that? Can‘t we use the container to ask for classes implementing a certain interface? The finder feels strange to me.

Reply:
There is no interface for fragments, same as there there is none for controllers in Symfony.

Reply:
We could introduce a marker interface like done in many places in Symfony. But we can also have the Controller directory hardcoded. Just needs to be recursive...

Reply:
I have introduced the recursion recently.